### PR TITLE
add test

### PR DESCRIPTION
**Pull request by drisspg**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132462
* #131844
* #131867
* #131863
* #131060

### CODE REVIEW

File: test/test_transformers.py [hunk 1]
```diff
@@ -2143,6 +2143,72 @@ def test_scaled_dot_product_fused_attention_with_inf(self, device):
...
+            torch.testing.assert_close(bg, mg, atol=3e-3, rtol=0)
+            self.assertTrue(not bg.isnan().any())
+            self.assertTrue(not mg.isnan().any())
+
+        # Check if masked rows are zero in output
+        mask_sum = mask.sum(dim=-1, keepdim=True)
+        masked_rows = (mask_sum == 0).expand_as(backend_out)
+        self.assertTrue((mask_sum == 0).sum() > 0, "No fully masked out rows found")
+        assert torch.all(backend_out[masked_rows] == 0), \
+            f"Non-zero values in fully masked rows for {backend}"
```
Comment (Skylion007) [inline test/test_transformers.py:2206]: Suggestion: use 'f"{backend=}"' to make failing asserts more readable (will print as backend=value)

**Review by Skylion007 — APPROVED** [review]

File: test/test_transformers.py [hunk 2]
```diff
@@ -2143,6 +2143,72 @@ def test_scaled_dot_product_fused_attention_with_inf(self, device):
...
+
+            # Create a mask with random row masking
+            mask = torch.ones(1, 1, seq_len, seq_len, dtype=torch.bool, device=device)
+
+            # Randomly select rows to mask
+            random_mask = torch.rand(seq_len, device=device) < mask_probability
+            mask[0, 0, random_mask, :] = False
+
+            # Randomly mask out some elements on along rows with 20% probz
+            mask = mask & (torch.rand(1, 1, seq_len, seq_len, device=device) < 0.2)
```
Comment (jbschlosser) [inline test/test_transformers.py:2165]: are we guaranteed to get a fully-masked out row with this setup? the randomness is making me question it

TBH I'd feel better with a really simple test that deterministically masks out a single row (I'm sure there's some repros from the infamous MHA + NaN issue)

File: test/test_transformers.py [hunk 3]
```diff
@@ -2143,6 +2143,72 @@ def test_scaled_dot_product_fused_attention_with_inf(self, device):
...
+
+        # Compare outputs
+        torch.testing.assert_close(backend_out, math_out, atol=5e-3, rtol=0)
+        self.assertTrue(not backend_out.isnan().any())
+        self.assertTrue(not math_out.isnan().any())
+        # Compare gradients
+        for bg, mg in zip(backend_grads, math_grads):
+            torch.testing.assert_close(bg, mg, atol=3e-3, rtol=0)
+            self.assertTrue(not bg.isnan().any())
+            self.assertTrue(not mg.isnan().any())
```
Comment (jbschlosser) [inline test/test_transformers.py:2199]: nit: could use `self.assertFalse` ?

File: test/test_transformers.py [hunk 4]
```diff
@@ -2143,6 +2143,72 @@ def test_scaled_dot_product_fused_attention_with_inf(self, device):
...
+
+        # Check if masked rows are zero in output
+        mask_sum = mask.sum(dim=-1, keepdim=True)
+        masked_rows = (mask_sum == 0).expand_as(backend_out)
+        self.assertTrue((mask_sum == 0).sum() > 0, "No fully masked out rows found")
```
Comment (jbschlosser) [inline test/test_transformers.py:2204]: oh lol you assert that we have at least one masked-out row. If randomness were changing, I'd be worried about this test being flaky. I'll refer back to my previous comment that a deterministic test with a guaranteed fully masked-out row might be preferred for maximum confidence

File: test/test_transformers.py [hunk 5]
```diff
@@ -2143,6 +2143,72 @@ def test_scaled_dot_product_fused_attention_with_inf(self, device):
...
+        if backend == SDPBackend.EFFICIENT_ATTENTION and "cpu" in str(device):
+            unittest.skip("EfficientAttention does not support masks on cpu")
+            return
+        query, key, value, mask = attention_inputs(seq_len, head_dim, device, dtype)
+
+        # Compute results for the tested backend
+        backend_out, backend_grads = compute_output_and_grads(query, key, value, mask, backend)
+
+        # Compute results for the Math backend
+        math_out, math_grads = compute_output_and_grads(query, key, value, mask, SDPBackend.MATH)
```
Comment (jbschlosser) [inline test/test_transformers.py:2189]: out of curiosity, since we're comparing against a math reference below, does this depend on the recent changes to use float32 precision for float16 math?

**Review by jbschlosser — APPROVED** [review]
Nice! it's great to have a test verifying the masked-out row behavior at this level. couple nits below; suggesting a simpler, deterministic test (i.e. guarantee a fully-masked out row) but it's not show-stopping

File: test/test_transformers.py (hunk 5 above)
Comment (drisspg) [reply test/test_transformers.py:2189]: The tolerances are the only thing that would rely on the updates to fp32 precision. This is more of a check that the updates to the fused kernels indeed reflects the new `_safe_softmax`

Comment (jbschlosser) [reply test/test_transformers.py:2192]: sweet. depending on how the NaN -> zero conversion story ends up, I'll propose a test with input NaNs (not from fully masked-out rows) comparing fused and math kernels to ensure consistency in how those are handled

**Comment by drisspg** [comment]
Closing for this stack: https://github.com/pytorch/pytorch/pull/131863
